compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp

```diff
@@ -690,13 +691,153 @@ FailureOr<Flow::WorkloadBuilder> getWorkloadBuilder(OpBuilder &builder,
+      return prePos.has_value() && nextPos.has_value() &&
+             (prePos.value() + 1) == nextPos;
+    });
+  };
+
+  // Finds that d2-d3 and d6-d7 are contiguous for the case (d1, d0, d2, d3, d5,
+  // d4, d6, d7).
+  ReassociationIndices range;
+  for (auto [idx, dim] : llvm::enumerate(loopDims)) {
+    if (!range.empty()) {
+      if (!areDimensionsContiguous(range.back(), dim) ||
+          (range.back() + 1 != dim)) {
+        if (range.size() > 1)
+          contiguousLoops.push_back({range.begin(), range.end()});
+        range.clear();
+      }
+    }
+    range.push_back(dim);
+  }
+  if (range.size() > 1) contiguousLoops.push_back(range);
+
+  return contiguousLoops;
+}
+
+/// Finds all collapsible loop ranges of type "parallel" without
+/// mixing them with "reduction" types.
+static SmallVector<ReassociationIndices> getCollapsibleLoops(
+    linalg::GenericOp genericOp) {
+  SmallVector<unsigned int> pDims;
+  genericOp.getParallelDims(pDims);
+
+  SmallVector<ReassociationIndices> pLoops =
+      getContiguousLoops(genericOp, pDims);
+
+  LLVM_DEBUG({
+    llvm::dbgs() << "Collapsing parallel loops: ";
+    for (auto indices : pLoops) {
+      llvm::dbgs() << "[";
+      for (auto idx : indices) llvm::dbgs() << idx << ",";
+      llvm::dbgs() << "]\t";
+    }
+    llvm::dbgs() << "\n";
+  });
+
+  return pLoops;
+}
+
+/// Collapse possible dimension of the given linalg.generic and return the new
+/// one
+static FailureOr<linalg::GenericOp> collapseLinalgGeneric(
+    TensorDimTrackingRewriter &rewriter, linalg::GenericOp genericOp) {
+  SmallVector<ReassociationIndices> collapseIndices =
+      getCollapsibleLoops(genericOp);
+
+  if (collapseIndices.empty()) return genericOp;
+
+  rewriter.setInsertionPoint(genericOp);
+  FailureOr<SmallVector<Value>> replacements =
+      mlir::linalg::collapseGenericOpIterationDims(genericOp, collapseIndices,
+                                                   rewriter);
+  if (failed(replacements) || replacements->empty()) {
+    return rewriter.notifyMatchFailure(genericOp,
+                                       "failed to collapse dimensions");
+  }
+
+  // Find and return collapsed linalg.generic
+  if (auto expandshapeOp =
+          replacements->front().getDefiningOp<tensor::ExpandShapeOp>()) {
+    if (auto newGenericOp =
+            expandshapeOp.getOperand().getDefiningOp<linalg::GenericOp>()) {
+      rewriter.replaceOp(genericOp, *replacements);
+      return newGenericOp;
+    }
+  }
+
+  return failure();
+}
+
+/// Returns true if the given op is collapsable.
+static bool isEligibleForCollapse(Operation *op,
+                                  ArrayRef<Operation *> producers) {
+  if (!producers.empty()) return false;
+
+  auto genericOp = dyn_cast<linalg::GenericOp>(op);
+  if (!genericOp) return false;
+
+  // TODO(guray) There is no mechanism to convey collapsed indexes from
+  // `tensor.collapse_shape` to `tensor.expand_shape`. Once we have this support
```

Comment:
`tensor.collapse_shape` doesnt need any additional support. Those already support inputs being fully dynamic. This is only needed for `tensor.expand_shape`.